New parser, bug fixes and new functionality.

collerek opened the pull request:
Hi, first of all thanks for an awesome job with the library! 
The rules you invented to classify given types of metadata (columns, tables etc.) were of great value!

Now I needed something like #78 so I had a go for it. 

## New Parser

In the process I read your #83 and #98 issues and they contain API that I like much more than the old one, so I decided to try implementing the new parser and iteration described in those issues along the way (more or less - i.e. you don't write how would you like to treat multi - token names (like `db.schema.table` etc.). So it's not exactly how you described it but it's close and I think this is the step in the right direction to get it how you see it in the future.


### Functionalities
Apart from implementing #83 and #98 the new functionalities are:

* Ability to get a dictionary containing column names divided by part of the query the column is located in (one of `select, where, order_by, join, insert and update`). Should close #78. (tested)
* Support for extracting with statement names so they are excluded from tables. (note that I corrected the sql examples as in most backends (all supported?) the with as subquery should be wrapped in parenthesis). Should close #112 (tested)
* Support for quoted parameters (basically removes them as in all supported backends they are used to wrap db identifiers, single quotes are used for strings). Should close #85 (tested)
* Introduce support for aliases without the `as` keyword. Should close #97 (in much simpler way) (tested)
* Introduce getting values list and values_dict (column, value pairs) from insert queries. Should close #103 (tested)
* Introduce a way to extract subqueries with their names in a dict
* Introduce a way to get subqueries names as a list
* Introduce a way to get columns without the ones coming from subqueries
* Introduce a way to extract inline comments as a list
* In the process changes and fixes causes also #120 to work correctly (tested)
* Parser properties are cached (in "protected" vars) for subsequent quicker access

### Bug fixes
Fixed yet unreported bugs:

*  Regex used to remove comments was greedy so if two (or more) comments were in a query it was matching also query text between the comments
* Fix wrongly excluded columns in update queries after SET (`update table set column = 'aa'`) etc.
* Fix wrongly classified as columns names of the subqueries

### Tests
All tests that were commented out or skipped now run and pass. I divided tests into smaller functions and multiple files to allow easier testing of selected sections of functionality if something fails. Coverage 100% and new tests for new functionalities.

So basically it should close all issues but #35 and at the same time introduces a lot of improvements etc.

## Open issues
Now I also want to introduce column aliases but got stuck here a little as some issues require clarification.

### Column aliases
As of now when you have a query like:
`Select column1 as myname from table` both `column1` and `myname` go into the columns, while they are the same column.
I believe aliases like this should be resolved back to the most base column, and aliases like `myname` should not end up in columns but I need your opinion on this. That also leaves cases like `select count(*) as counter`, should `counter` be included in columns?

#### Motivation
My use case is that I need all real (and only real) database columns used in a parsed query to later reflect used databases and check column types. That means that (in perfect world) I would get: 

* recursively resolve column names so `select col1 as myname from table` should return only `table.col1` in columns
* recursively resolve names of columns from subqueries so `select sub.test1 from (select aa as test1 from table) sub` should return columns = ["table.aa"] as sub.test1 is a subquery alias and test1 is alias of aa.
* try to assign table for each column even if prefix is not set so: `select aa.col1, col2 from table aa` should return ["table.col1", "table.col2"] and there should be a way of returning column with table in a dict. Of course if  table cannot be resolved (like select aa, bb from tab1, tab2` where you cannot tell which column is from which table both column should return a list of ["tab1", "tab2"]

I can already get and extract subqueries but would need column aliases (and later optionally best match assigned tables to columns). But for that I need confirmation from you that I think right and aliases of columns should not end up in columns. Or alternatively there should be a way to get only aliases (as they are in end result of the query i.e. in select) or only the real database columns.


In my opinion the end product is much more concise, cleaner and easier to maintain, but it's basically re-written from scratch so let me know what you think! :) I updated readme so best to start there, rest is in tests since you do not have documentation.

----

Resolves #103, resolves #83, resolves #98, resolves #78 and resolves #120.

collerek commented on pyproject.toml:
```diff
@@ -15,6 +15,7 @@ packages = [
 [tool.poetry.dependencies]
 python = "^3.6"
 sqlparse = "^0.4.1"
+black = "^20.8b1"
```
Probably my lack of experience with poetry 🤣 didn't touch pyproject but had problems with installing black so maybe that's what added the duplicate

macbre replied:
No bother 🙂

macbre replied:
@collerek - can you resolve `poetry.lock` file conflict? You can simply copy `poetry.lock` and `pyproject.toml` from master branch. There's not need to install `black` dependency in two places.

collerek replied:
Removed double

macbre commented:
> My use case is that I need all real (and only real) database columns used in a parsed query to later reflect used databases and check column types. That means that (in perfect world) I would get:

That's my idea behind aliases handling too. I'd like `sql-metadat` to return real columns that are used in the query, to allow other tools to analyse which columns from the real tables are used.

Hence, `select count(*) as counter` should not add any column to the list - `counter` is an alias of an aggregate function.

It's a more complex topic, so let's move the discussion to a different PR / issue and continue with the refactoring / new API here.

> In my opinion the end product is much more concise, cleaner and easier to maintain, but it's basically re-written from scratch so let me know what you think! :) I updated readme so best to start there, rest is in tests since you do not have documentation.

I do agree! Thank you very much!

macbre commented:
I've run pytest locally on this PR branch and I think there's a typo in assertions for:

```
FAILED test/test_getting_tables.py::test_db2_query - AssertionError: assert ['CUSTOMER_ACCOUNT.IDENTIFICATION_CODE',\n 'identificationCode',\n 'economic_operator.KBO_NUMBER',\n 'kboNumber',\n 'economic_operat...
```

`pytest -vv` gives me the following diff:

```diff
E         -  'ca1_remainder_t    otal_guaranteed.accounting_entity_id',
E         ?                  ----
E         +  'ca1_remainder_total_guaranteed.accounting_entity_id',
```

A tab inserted there? 🙂

macbre commented on test/test_getting_tables.py:
```diff
@@ -0,0 +1,404 @@
+        "contact_details.id",
+        "CUSTOMER_ACCOUNT.contact_details_id",
+        "accounting_entity.id",
+        "ca1_remainder_t    otal_guaranteed.accounting_entity_id",
```
```suggestion
        "ca1_remainder_total_guaranteed.accounting_entity_id",
```

macbre commented on sql_metadata/generalizator.py:
```diff
@@ -0,0 +1,104 @@
+    @property
+    def comments(self) -> List[str]:
+        """
+        Removes comments from SQL query
```
```suggestion
        Gets comments from SQL query
```

macbre commented on sql_metadata/parser.py:
```diff
@@ -0,0 +1,533 @@
+        """
+        Resets cached attributes before setting new query
+        """
+        self._tokens = None
+
+        self._columns = None
+        self._columns_dict = None
+
+        self._tables = None
+        self._table_aliases = None
+
+        self._with_names = None
+        self._subqueries = None
+        self._subqueries_names = None
+
+        self._limit_and_offset = None
+
+        self._values = None
+        self._values_dict = None
+
+        self._raw_query = value
```
I'm wondering if this is needed. Is there any scenario where `self._query` will change? I believe `_query` is only set (and pre-processed) when `Parser` instance is created.

collerek replied:
You still can manually assign Parser.query = "New sql" and either we need to raise exception (not allowed, use new instance or something), or reset the cached props as they will still have values from old query and will return them instead of running

macbre replied:
I'd rather not allow people to do that. Let's assume that `Parser` instance is run for a specific query. Want to parse another query - create a new instance....

collerek replied:
I think just removing query.setter should do the trick

macbre replied:
Agree. Let's do that 👌🏻

collerek replied:
Removed.

macbre commented on sql_metadata/token.py:
```diff
@@ -0,0 +1,162 @@
+        """
+        Representation - useful for debugging
+        """
+        repr_str = ["=".join([str(k), str(v)]) for k, v in self.__dict__.items()]
+        return f"SQLToken({','.join(repr_str)})"
```
👌🏻

macbre commented on sql_metadata/parser.py:
```diff
@@ -0,0 +1,533 @@
+        open_parenthesises = []
+        for index, tok in enumerate(non_empty_tokens):
+            token = SQLToken(
+                value=tok.value,
+                is_keyword=tok.is_keyword,
+                is_name=tok.ttype is Name,
+                is_punctuation=tok.ttype is Punctuation,
+                is_dot=str(tok) == ".",
+                is_wildcard=tok.ttype is Wildcard,
+                is_integer=tok.ttype is Number.Integer,
+                is_float=tok.ttype is Number.Float,
+                is_left_parenthesis=str(tok) == "(",
+                is_right_parenthesis=str(tok) == ")",
```
I'm wondering if we shouldn't rather pass `tok` into the constructor of SQLToken to move this logic into the token module. Parser is already quite complex 🙂

collerek replied:
Yeah sounds like a good idea :)

collerek replied:
Maybe I will wait for your further review to not change anything major while you are still in the process. ;)

macbre replied:
Sure, sure. Let's tackle this one is another PR. And limit the scope of this one. It's already quite big, isn't? 😉

collerek replied:
Yeah, sorry for that :D it was like some changes caused new bugs or I found a functionality that  and the scope exploded a little 😅

macbre commented on sql_metadata/keywords_lists.py:
```diff
@@ -0,0 +1,82 @@
+    "NOW",
+    "LAST_DAY",
+    "DATE_ADD",
+    "COALESCE"
```
```suggestion
    "COALESCE",
```

`black` tells me:

```
reformatted sql_metadata/keywords_lists.py
All done! ✨ 🍰 ✨
1 file reformatted, 23 files left unchanged.
```

macbre commented on test/test.sql:
```diff
@@ -0,0 +1,215 @@
+
+WHERE 1 = 1
+AND C.START_OF_VALIDITY <= DATE'2020-01-01'
+AND C.END_OF_VALIDITY > DATE'2020-01-01'
```
Good luck refactoring this ~query~ Query 😁

collerek replied:
That's coming straight from #97, to be honest I didn't read it in depth, just compared the end result to expected output in tables aliases and tables 🙉

collerek replied:
That's the problem with examples coming from users - they tend to be not valid sql (wrong quotes, missing parenthesis etc), I don't know db2 dialect, but just checked and it should be date with () so will fix that

macbre replied:
Makes sense.

collerek replied:
Added () to date, there is still weird & (variable) but not sure how that should be handled

macbre commented on test/test_caching.py:
```diff
@@ -0,0 +1,9 @@
+    parser = Parser("Select * from test")
+    assert parser.tables == ["test"]
+
+    parser.query = "Select * from test2"
```
Now let's check that such assignment raises an exception.

collerek replied:
Applied.

macbre commented on test/test_complex_aliases.py:
```diff
@@ -0,0 +1,60 @@
+    with open(sql_filename, "r", encoding="latin-1") as content_file:
+        content = content_file.read()
+    content = sqlparse.format(content, strip_comments=True).strip()
+    parser = Parser(content)
```
Maybe we can improve the `Parser` and allow it to consume either strings or file object?

macbre commented on test/test_complex_aliases.py:
```diff
@@ -0,0 +1,60 @@
+    sql_filename = f"{dir_path}/test.sql"
+    with open(sql_filename, "r", encoding="latin-1") as content_file:
+        content = content_file.read()
+    content = sqlparse.format(content, strip_comments=True).strip()
```
Why is this step necessary?

collerek replied:
Not sure, took it directly from PR #99, I guess it should work without it - will check

collerek replied:
BTW. maybe we should use sqlparse functionality instead of regex to remove comments if it already exist?

collerek replied:
Confirm it works just fine without the sqlparse line

macbre replied:
That's a good idea - query pre-parsing step 👌🏻

collerek replied:
Removed extra step.

macbre commented on test/test_caching.py:
```diff
@@ -1,9 +1,13 @@
-    parser.query = "Select * from test2"
-    assert parser._tables is None
+    with pytest.raises(AttributeError):
+        parser.query = "Select * from test2"
```
👌🏻

macbre commented on .github/workflows/python-ci.yml:
```diff
@@ -60,7 +60,7 @@ jobs:
 
     - name: Lint with pylint
-      run: poetry run pylint sql_metadata.py
+      run: poetry run pylint sql_metadata
```
Let's change `pyproject.toml` too:

```python
packages = [
    { include="sql_metadata.py" }
]
```

Right now CI fails with (there's no `sql_metadata.py` file anymore):

```
Run poetry install
Creating virtualenv sql-metadata in /home/runner/work/sql-metadata/sql-metadata/.venv
Installing dependencies from lock file

Package operations: 23 installs, 0 updates, 0 removals

...

  ValueError

  /home/runner/work/sql-metadata/sql-metadata/sql_metadata.py does not contain any element

  at ~/.poetry/lib/poetry/_vendor/py3.9/poetry/core/masonry/utils/package_include.py:60 in check_elements
      56│         return any(element.suffix == ".py" for element in self.elements)
      57│ 
      58│     def check_elements(self):  # type: () -> PackageInclude
      59│         if not self._elements:
    → 60│             raise ValueError(
      61│                 "{} does not contain any element".format(self._base / self._include)
      62│             )
      63│ 
      64│         root = self._elements[0]
```

collerek replied:
Changed

macbre commented on sql_metadata/generalizator.py:
```diff
@@ -56,7 +58,7 @@ def remove_comments(self) -> str:
         :rtype: str
         """
-        sql = re.sub(r"\s?/\*.+?\*/", "", self._raw_query)
+        sql = sqlparse.format(self._raw_query, strip_comments=True)
```
👌🏻

macbre commented on sql_metadata/generalizator.py:
```diff
@@ -0,0 +1,106 @@
+    Class used to produce generalized sql out of given query
+    """
+
+    def __init__(self, sql: Optional[str] = None):
```
Why is `sql` argument optional? What's the point of passing `None` here?

collerek replied:
In practice probably not very useful 🙈 
Since you can: `assert Parser("").tokens == []`

I assumed that that one should also pass: `assert Parser().tokens == []`

Like if you get the queries from somewhere and even if the query is None you can parse them all in a loop but that's a little far stretched 🤷

macbre replied:
Then let's default to an empty string instead of `None`:

```suggestion
    def __init__(self, sql: str = ""):
```

macbre commented on sql_metadata/generalizator.py:
```diff
@@ -0,0 +1,106 @@
+        comments = re.findall(r"\s?/\*.+?\*/", self._raw_query)
+        return [x.strip() for x in comments]
+
+    @property
+    def remove_comments(self) -> str:
```
I'd rename it to be a noun as it;s a property after all.

So either

```suggestion
    @property
    def without_comments(self) -> str:
```

or

```suggestion
    def remove_comments(self) -> str:
```

collerek replied:
How do you prefer?

collerek replied:
Since most/all of them are properties maybe the first one to be in line with most features?

macbre replied:
ea02fa1

macbre commented:
Tests are green now - we're on good path 🙂

collerek commented:
You want me to change the names or you will fix that?

macbre commented:
> You want me to change the names or you will fix that?

Please do so.

macbre commented:
Wow! That was quite a journey with code reviewing your changes, @collerek. Let's address the remaining open comments and we're ready to merge this one.

Once again - dzięki 🙂 🇵🇱

collerek commented:
Hope that's not due to the crappy quality of my code 🤣 

Polecam się na przyszłość i dzięki za super libkę 😉 🇵🇱

macbre commented:
> Hope that's not due to the crappy quality of my code

Not at all.

macbre commented:
Resolves #103, #83, #98, #78, #120.

collerek commented:
Any reason you didn't mark #112, #85  and #97 as resolved? 
Need more tests?

Let me know what are the next steps as I would like to help with column aliases if I can as this is my final goal :)

macbre commented:
> Any reason you didn't mark #112, #85 and #97 as resolved? Need more tests?

I just added this small comment to mark these additional issues as resolved once this PR is merged (via "resolved #..." GH feature). All is good. Sorry for the confusion here.

macbre commented:
> Let me know what are the next steps as I would like to help with column aliases if I can as this is my final goal :)

Let's tackle aliases in the next PR, ok? I'll complete the review and merge this one tomorrow.

collerek commented:
> Let's tackle aliases in the next PR, ok? I'll complete the review and merge this one tomorrow.

Sure, that's what I had in mind, saw that you added more issues to v.2.0 milestone that's why I'm asking if you are merging this one or wait for some other stuff. 

We still have two new issues coming from this one (ability to pass file (or file path?) and refactoring token logic into token init, that's why I'm asking what is the plan, no rush or whatever, just curious :)

macbre commented on test/test_normalization.py:
```diff
@@ -1,79 +1,79 @@
-def test_generalize_sql():
-    assert generalize_sql(None) is None
+def test_generalization_of_sql():
+    assert Parser(None).generalize is None
```
```suggestion
    assert Parser("").generalize == ""
```

`None` is no longer support here.

collerek commented:
I just realized that removing comments in _preprocess in init is not really necessary and since we use sqlparse now in remove_comments it parses the query two times which hits performance, maybe we should revert and remove the step if it's working without it anyway (more tokens in parsed query with comments but I think it's still quicker than removing comments)?

Can you check performance on your machine? (as it is now and without step 3 in preprocess)

macbre commented on sql_metadata/parser.py:
```diff
@@ -0,0 +1,508 @@
+        self._columns_dict = self._columns_dict or dict()
+        self._columns_dict.setdefault(section, UniqueList()).append(column)
+
+    def _preprocess_query(self) -> Optional[str]:
+        """
+        Perform initial query cleanup
+
+        :rtype str
+        """
+        # 0. remove newlines
+        if self._raw_query is None:
+            return None
```
`self._raw_query` can no longer by `None`.

```suggestion
    def _preprocess_query(self) -> str:
        """
        Perform initial query cleanup

        :rtype str
        """
        if self._raw_query == "":
            return ""

        # 0. remove newlines
```

So just leave the empty string check to leave early.

macbre replied:
And we may simply pass the SQL query to preprocess as an argument here, making this method a static one. This will make it easy to unit test this method.

But I'll deal with that later on - let's merge it as it is now 🙂

macbre commented:
> I just realized that removing comments in _preprocess in init is not really necessary and since we use sqlparse now in remove_comments it parses the query two times which hits performance, maybe we should revert and remove the step if it's working without it anyway (more tokens in parsed query with comments but I think it's still quicker than removing comments)?

Sure, please do that and add a test case in `test_preprocessing` with a query with a comment. This will make sure that we keep the behaviour of pre-processing consistent and don't break our contract.

macbre commented on test/test_query.py:
```diff
@@ -30,6 +30,12 @@ def test_preprocessing():
+        Parser("SELECT /*my random comment*/ foo, id FROM `db`.`test`").query
+        == "SELECT /*my random comment*/ foo, id FROM db.test"
+    )
+
```
👌🏻

collerek commented:
> > I just realized that removing comments in _preprocess in init is not really necessary and since we use sqlparse now in remove_comments it parses the query two times which hits performance, maybe we should revert and remove the step if it's working without it anyway (more tokens in parsed query with comments but I think it's still quicker than removing comments)?
> 
> Sure, please do that and add a test case in `test_preprocessing` with a query with a comment. This will make sure that we keep the behaviour of pre-processing consistent and don't break our contract.
> 
> `sql-metadata` user can still call `parser.without_comments` to get the query without comments.

I reverted the step 3 and added test in preprocess but since we don't remove it there is not much to test 😅 
I added this step in one of the recent commits with switching to sqlparse, it was not there in old parser and in new parser just few commits back.

macbre commented on test/test_query.py:
```diff
@@ -1,283 +1,60 @@
-    assert ["tab", "tab2"] == get_query_tables(
-        """select a,b,c from tab full  outer \r\n\t  join tab2  on (col1 = col2) group   
-               \r\n   \t   by  a, b, c """
+    # normalize newlines
+    assert (
+        Parser("SELECT /*my random comment*/ foo, id FROM `db`.`test`").query
+        == "SELECT /*my random comment*/ foo, id FROM db.test"
     )
```
```suggestion
    # comments are kept
    assert (
        Parser("SELECT /*my random comment*/ foo, id FROM `db`.`test`").query
        == "SELECT /*my random comment*/ foo, id FROM db.test"
    )
```

macbre commented:
We're ready to merge once CI passes 🙂 Thanks for your great work here, @collerek.

Looking forward to your next PRs and ideas. Have a good one.
